morl_baselines/multi_policy/pcn/pcn.py

```diff
@@ -0,0 +1,441 @@
+"""Pareto Conditioned Network. Code adapted from https://github.com/mathieu-reymond/pareto-conditioned-networks ."""
+import heapq
+import os
+from dataclasses import dataclass
+from typing import Optional, Union
+
+import gymnasium as gym
+import numpy as np
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+
+from morl_baselines.common.morl_algorithm import MOAgent, MOPolicy
+from morl_baselines.common.performance_indicators import (
+    expected_utility,
+    hypervolume,
+    sparsity,
+)
+from morl_baselines.common.utils import random_weights
+
+
+def crowding_distance(points):
+    """Compute the crowding distance of a set of points."""
+    # first normalize across dimensions
+    points = (points - points.min(axis=0)) / (points.ptp(axis=0) + 1e-8)
+    # sort points per dimension
+    dim_sorted = np.argsort(points, axis=0)
+    point_sorted = np.take_along_axis(points, dim_sorted, axis=0)
+    # compute distances between lower and higher point
+    distances = np.abs(point_sorted[:-2] - point_sorted[2:])
+    # pad extrema's with 1, for each dimension
+    distances = np.pad(distances, ((1,), (0,)), constant_values=1)
+    # sum distances of each dimension of the same point
+    crowding = np.zeros(points.shape)
+    crowding[dim_sorted, np.arange(points.shape[-1])] = distances
+    crowding = np.sum(crowding, axis=-1)
+    return crowding
+
+
+@dataclass
+class Transition:
+    """Transition dataclass."""
+
+    observation: np.ndarray
+    action: int
+    reward: np.ndarray
+    next_observation: np.ndarray
+    terminal: bool
+
+
+def get_non_dominated(solutions):
+    """Returns the non-dominated solutions from a set of solutions."""
+    is_efficient = np.ones(solutions.shape[0], dtype=bool)
+    for i, c in enumerate(solutions):
+        if is_efficient[i]:
+            # Remove dominated points, will also remove itself
+            is_efficient[is_efficient] = np.any(solutions[is_efficient] > c, axis=1)
+            # keep this solution as non-dominated
+            is_efficient[i] = 1
+    return is_efficient
+
+
+def compute_hypervolume(q_set, ref):
+    """Computes hypervolume of the q_set."""
+    nA = len(q_set)
+    q_values = np.zeros(nA)
```

Comment:
Then this should be named hvs

Reply:
this function was never used, removed it